File: samples/scala-replicatedentity-examples/src/main/scala/com/example/replicated/map/domain/SomeMap.scala
package com.example.replicated.map.domain

import com.akkaserverless.replicatedentity.ReplicatedData
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedEntity
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedEntityContext
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedMap
import com.example.replicated.map
import com.google.protobuf.empty.Empty
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedRegister
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedSet

// This class was initially generated based on the .proto definition by Akka Serverless tooling.
//
// As long as this file exists it will not be overwritten: you can maintain it yourself,
// or delete it so it is regenerated as needed.

/** A replicated entity. */
class SomeMap(context: ReplicatedEntityContext) extends AbstractSomeMap {

  // keys
  private val FooKey = SomeKey("foo")
  private val BarKey = SomeKey("bar")
  private val BazKey = SomeKey("baz")

  /** Command handler for "IncreaseFoo". */
  def increaseFoo(currentData: ReplicatedMap[SomeKey, ReplicatedData], increaseFooValue: map.IncreaseFooValue): ReplicatedEntity.Effect[Empty] = {
    val foo = currentData.getReplicatedCounter(FooKey)
    effects
      .update(currentData.update(FooKey, foo.increment(increaseFooValue.value)))
      .thenReply(Empty.defaultInstance)
  }

  /** Command handler for "DecreaseFoo". */
  def decreaseFoo(currentData: ReplicatedMap[SomeKey, ReplicatedData], decreaseFooValue: map.DecreaseFooValue): ReplicatedEntity.Effect[Empty] = {
    val foo = currentData.getReplicatedCounter(FooKey)
    effects
      .update(currentData.update(FooKey, foo.decrement(decreaseFooValue.value)))
      .thenReply(Empty.defaultInstance)
  }

  /** Command handler for "SetBar". */
  def setBar(currentData: ReplicatedMap[SomeKey, ReplicatedData], setBarValue: map.SetBarValue): ReplicatedEntity.Effect[Empty] = {
    val bar: ReplicatedRegister[String] = currentData.getReplicatedRegister(BarKey)
    effects
      .update(currentData.update(BarKey, bar.set(setBarValue.value)))
      .thenReply(Empty.defaultInstance)
  }

  /** Command handler for "AddBaz". */
  def addBaz(currentData: ReplicatedMap[SomeKey, ReplicatedData], addBazValue: map.AddBazValue): ReplicatedEntity.Effect[Empty] = {
    val baz: ReplicatedSet[String] = currentData.getReplicatedSet(BazKey)
    effects
      .update(currentData.update(BarKey, baz.add(addBazValue.value)))
      .thenReply(Empty.defaultInstance)
  }

  /** Command handler for "RemoveBaz". */
  def removeBaz(currentData: ReplicatedMap[SomeKey, ReplicatedData], removeBazValue: map.RemoveBazValue): ReplicatedEntity.Effect[Empty] = {
val baz: ReplicatedSet[String] = currentData.getReplicatedSet(BazKey)
    effects
      .update(currentData.update(BarKey, baz.remove(removeBazValue.value)))
      .thenReply(Empty.defaultInstance)
  }

  /** Command handler for "Get". */
  def get(currentData: ReplicatedMap[SomeKey, ReplicatedData], getValues: map.GetValues): ReplicatedEntity.Effect[map.CurrentValues] = {

    val foo = currentData.getReplicatedCounter(FooKey)
    val bar: ReplicatedRegister[String] = currentData.getReplicatedRegister(BarKey, () => "")
    val baz: ReplicatedSet[String] = currentData.getReplicatedSet(BazKey)

    val resp = map.CurrentValues(foo.value, bar(), baz.elements.toSeq)
    effects.reply(resp)
  }
    
}

File: samples/java-protobuf-replicatedentity-examples/src/main/java/com/example/replicated/map/domain/SomeMap.java
package com.example.replicated.map.domain;

import kalix.javasdk.replicatedentity.ReplicatedCounter;
import kalix.replicatedentity.ReplicatedData;
import kalix.javasdk.replicatedentity.ReplicatedEntityContext;
import kalix.javasdk.replicatedentity.ReplicatedMap;
import kalix.javasdk.replicatedentity.ReplicatedRegister;
import kalix.javasdk.replicatedentity.ReplicatedSet;
import com.example.replicated.map.SomeMapApi;
import com.google.protobuf.Empty;

import java.util.stream.Collectors;

public class SomeMap extends AbstractSomeMap {
  @SuppressWarnings("unused")
  private final String entityId;

  private static final SomeMapDomain.SomeKey FOO_KEY =
      SomeMapDomain.SomeKey.newBuilder().setSomeField("foo").build();

  private static final SomeMapDomain.SomeKey BAR_KEY =
      SomeMapDomain.SomeKey.newBuilder().setSomeField("bar").build();

  private static final SomeMapDomain.SomeKey BAZ_KEY =
      SomeMapDomain.SomeKey.newBuilder().setSomeField("baz").build();

  public SomeMap(ReplicatedEntityContext context) {
    this.entityId = context.entityId();
  }

  // tag::update[]
  @Override
  public Effect<Empty> increaseFoo(
      ReplicatedMap<SomeMapDomain.SomeKey, ReplicatedData> map,
      SomeMapApi.IncreaseFooValue command) {
    ReplicatedCounter foo = map.getReplicatedCounter(FOO_KEY); // <1>
    return effects()
        .update(map.update(FOO_KEY, foo.increment(command.getValue()))) // <2> <3>
        .thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<Empty> decreaseFoo(
      ReplicatedMap<SomeMapDomain.SomeKey, ReplicatedData> map,
      SomeMapApi.DecreaseFooValue command) {
    ReplicatedCounter foo = map.getReplicatedCounter(FOO_KEY); // <1>
    return effects()
        .update(map.update(FOO_KEY, foo.decrement(command.getValue()))) // <2> <3>
        .thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<Empty> setBar(
      ReplicatedMap<SomeMapDomain.SomeKey, ReplicatedData> map, SomeMapApi.SetBarValue command) {
    ReplicatedRegister<String> bar = map.getReplicatedRegister(BAR_KEY); // <1>
    return effects()
        .update(map.update(BAR_KEY, bar.set(command.getValue()))) // <2> <3>
        .thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<Empty> addBaz(
      ReplicatedMap<SomeMapDomain.SomeKey, ReplicatedData> map, SomeMapApi.AddBazValue command) {
    ReplicatedSet<String> baz = map.getReplicatedSet(BAZ_KEY); // <1>
    return effects()
        .update(map.update(BAZ_KEY, baz.add(command.getValue()))) // <2> <3>
        .thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<Empty> removeBaz(
      ReplicatedMap<SomeMapDomain.SomeKey, ReplicatedData> map, SomeMapApi.RemoveBazValue command) {
    ReplicatedSet<String> baz = map.getReplicatedSet(BAZ_KEY); // <1>
    return effects()
        .update(map.update(BAZ_KEY, baz.remove(command.getValue()))) // <2> <3>
        .thenReply(Empty.getDefaultInstance());
  }
  // end::update[]

  // tag::get[]
  @Override
  public Effect<SomeMapApi.CurrentValues> get(
      ReplicatedMap<SomeMapDomain.SomeKey, ReplicatedData> map, SomeMapApi.GetValues command) {
    ReplicatedCounter foo = map.getReplicatedCounter(FOO_KEY); // <1>
    ReplicatedRegister<String> bar = map.getReplicatedRegister(BAR_KEY, () -> ""); // <1>
    ReplicatedSet<String> baz = map.getReplicatedSet(BAZ_KEY); // <1>
    SomeMapApi.CurrentValues currentValues =
        SomeMapApi.CurrentValues.newBuilder()
            .setFoo(foo.getValue())
            .setBar(bar.get())
            .addAllBaz(baz.elements().stream().sorted().collect(Collectors.toList()))
            .build();
    return effects().reply(currentValues);
  }
  // end::get[]
}
